### PR TITLE
scripts: runners: nrfjprog: Multiple improvements

### LINKED ISSUE #55340: nrfjprog runner only supports extra parameters for program commands
The nrfjprog runner only supports extra parameters (passed by "--tool-opt") for program commands, but ignores it for others like reset and recover.
This leads to an error like shown below when trying to access remote adapters (which are not officially supported by nrfjprog, as it seems, but anyway...):

```console
root@a6fdc28834bc:/workdir/app/nrf_netcore# west flash --skip-rebuild -r nrfjprog --dev-id 1050057458 --erase --tool-opt="--ip 192.168.56.1"
-- west flash: using runner nrfjprog
-- runners.nrfjprog: mass erase requested
-- runners.nrfjprog: Flashing file: /workdir/app/nrf_netcore/build/zephyr/zephyr.hex
[ #################### ]   0.736s | Erase file - Done erasing                                                          
[ #################### ]   0.892s | Program file - Done programming                                                    
[ #################### ]   0.874s | Verify file - Done verifying                                                       
ERROR: No debuggers were discovered.
NOTE: For additional output, try running again with logging enabled (--log).
NOTE: Any generated log error messages will be displayed.
FATAL ERROR: command exited with status 41: nrfjprog --pinreset -f NRF53 --snr 1050057458
```

When "tool-opt" parameters are taken into account for all commands, remote debuggers (connected via Segger J-Link Remote Server) can successfully be used for erasing, programming and verifying nRF boards (like my nRF5340-DK):

```console
root@a6fdc28834bc:/workdir/app/nrf_netcore# west flash --skip-rebuild -r nrfjprog --dev-id 1050057458 --erase --tool-opt="--ip 192.168.56.1"
-- west flash: using runner nrfjprog
-- runners.nrfjprog: mass erase requested
-- runners.nrfjprog: Flashing file: /workdir/app/nrf_netcore/build/zephyr/zephyr.hex
[ #################### ]   0.775s | Erase file - Done erasing                                                          
[ #################### ]   0.899s | Program file - Done programming                                                    
[ #################### ]   0.891s | Verify file - Done verifying                                                       
Applying pin reset.
-- runners.nrfjprog: Board with serial number 1050057458 flashed successfully.
```

These are the changes I applied to [nrfjprog.py](https://github.com/zephyrproject-rtos/zephyr/blob/a2f203d/scripts/west_commands/runners/nrfjprog.py):

```diff
--- a/scripts/west_commands/runners/nrfjprog.py
+++ b/scripts/west_commands/runners/nrfjprog.py
@@ -205,10 +205,10 @@ class NrfJprogBinaryRunner(ZephyrBinaryRunner):
         if self.family == 'NRF53':
             self.check_call(['nrfjprog', '--recover', '-f', self.family,
                              '--coprocessor', 'CP_NETWORK',
-                             '--snr', self.dev_id])
+                             '--snr', self.dev_id] + self.tool_opt)
 
         self.check_call(['nrfjprog', '--recover',  '-f', self.family,
-                         '--snr', self.dev_id])
+                         '--snr', self.dev_id] + self.tool_opt)
 
     def program_hex(self):
         # Get the nrfjprog command use to actually program self.hex_.
@@ -354,14 +354,14 @@ class NrfJprogBinaryRunner(ZephyrBinaryRunner):
     def reset_target(self):
         if self.family == 'NRF52' and not self.softreset:
             self.check_call(['nrfjprog', '--pinresetenable', '-f', self.family,
-                             '--snr', self.dev_id])  # Enable pin reset
+                             '--snr', self.dev_id] + self.tool_opt)  # Enable pin reset
 
         if self.softreset:
             self.check_call(['nrfjprog', '--reset', '-f', self.family,
-                             '--snr', self.dev_id])
+                             '--snr', self.dev_id] + self.tool_opt)
         else:
             self.check_call(['nrfjprog', '--pinreset', '-f', self.family,
-                             '--snr', self.dev_id])
+                             '--snr', self.dev_id] + self.tool_opt)
 
     def do_run(self, command, **kwargs):
         self.require('nrfjprog')

```

### PR DESCRIPTION
**Pull request by carlescufi**
```
    scripts: runners: nrfjprog: Use --tool-opt in all operations

    Pass on the tool-specific options to nrfjprog during all operations, and
    not only when programming. This is useful when combined with options
    like --ip, that allows west flash to be used over the network.

    Fixes #55340.

    Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
```
```
    scripts: runners: tests: nrfjprog: Test --tool-opt

    Add tests for the --tool-opt command-line switch, to ensure that if the
    user includes additional tool options those get passed on to nrfjprog at
    the end of the command-line to override any preceding ones.

    Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
```
```
    scripts: runners: nrf: Add support for multiple nRF backends

    Generalize the logic that was previously in nrfjprog.py into a new
    nrf_common.py, which can then use specific tool subclasses, one of which
    is nrfjprog.py, to implement the actual tool invocation.

    This commit lays down the groundwork for the addition of a new backend
    for nRF boards, the new nrfutil tool from Nordic. Both nrfjprog and
    nrfutil will coexist in the immediate future, but nrfutil is expected to
    be the sole tool for everything nRF in due time.  This is why the internal
    representation of the operations is based in JSON and is using the exact
    format that nrfutil expects when using the 'execute-batch' option, that
    takes a JSON file with an array of operations to be executed.

    Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
```
```
    scripts: west: runners: Support class hierarchies in the runners

    Allow for multiple levels of inheritance in the runners in order to make
    it possible to share common infrastructure in similar runners.
    Also check if the class can be instantiated (i.e. it is not abstract) to
    avoid returning abstract base classes.

    Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
```

Fixes #55340.

### CODE REVIEW

File: scripts/west_commands/runners/nrf_common.py [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+    def op_program(self, hex_file, erase, qspi_erase, defer=False, core=None):
+        args = {'firmware': {'file': hex_file, 'format': 'NRFDL_FW_INTEL_HEX'},
+                'chip_erase_mode': erase, 'verify': 'VERIFY_READ'}
+        if qspi_erase:
+            args['firmware']['qspi_erase_mode'] = qspi_erase
+        self.exec_op('program', defer, core, **args)
+
+    def exec_op(self, op, defer=False, core=None, **kwargs):
+        _op = f'{op}'
+        op = {'operation': {'type': _op}}
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/nrf_common.py:350]: `op = {'operation': {'type': f'{op}'}}`

**Review by mbolivar-nordic — APPROVED** [review]
LGTM; one nonblocking suggestion. Thanks for taking this on and making the `--tool-opt` improvements.

**Review by fabiobaltieri — APPROVED** [review]

File: scripts/west_commands/runners/nrf_common.py (hunk 1 above)
Comment (carlescufi) [reply scripts/west_commands/runners/nrf_common.py:350]: Sure, but why do you think it's better?

Comment (carlescufi) [reply scripts/west_commands/runners/nrf_common.py:350]: Oh I see, I no longer have the need for `_op` later on! I used to in an earlier version! Thanks!

**Pull request merged**